doc/arch/014-reporting-etl-pipeline.md

```diff
@@ -0,0 +1,150 @@
+occurrence in data warehouse design ([Here's one such example](https://www.mssqltips.com/sqlservertip/4054/creating-a-date-dimension-or-calendar-table-in-sql-server/)).
+
+The important elements of our calendar will be:
+* Cardinal month (1-12)
+* Year
+* Quarter: Quarterly analysis is common, so precomputing will be beneficial.
+* A unique date, like first of the month: This helps uniquely identify or group by each record in the calendar table on
+  a single column. `GROUP BY (year, month)` is more awkward.
+
+The calendar table will usually be joined with Simple's application tables to populate the denormalized reporting
+tables. The calendar table will be implemented as a non-materialized database view, since it is fairly small and not
+expensive to compute when needed.
+
+### Reports
+
+Since reporting queries will be simplified, we can start using Metabase for new reports. Metabase is ideal for quick
+prototyping and validation of reports with stakeholders because:
+
+* Easy-to-use interface
+* Easy to grant access to stakeholders, and controlled access to ourselves if necessary
+* Dynamic queries (eg. pick your own date range)
+
+Once reports are validated, we may want to move them in-house into the Simple Server application/dashboard. The benefits
+of doing so would be:
+
+* Better test coverage
+* Less fragmentation
+
+In any case, the long-term location of the reports themselves will not greatly impact the reporting table design
+described here.
+
+## Alternatives and constraints
+
+Apart from the proposed approach of reporting tables as database views, there are some alternatives that were
+considered.
+
+### Reporting tables in a separate database
+
+We could continue to generate the reporting tables by hand, but place them in a separate database. This would greatly
+alleviate performance problems. However, database performance is a secondary concern. Furthermore, the proposed approach
+will still set us up nicely to extract the reporting tables to a separate database in the future if necessary.
+
+### Rollup summary tables
+
+Rollup summary tables were proposed in [an earlier ADR](011-rollup-summary-tables-for-dashboards.md). The crux of the
+approach is to incrementally write data to these tables as application transactions occur. Eg. update a few rows in the
+"latest BPs per month" table when a new BP is recorded. The main drawback of this approach is that it's tough to iterate
+on the schemas, since every change would require a backfill of data from the beginning of time.
+
+### Third-party data warehouse like Looker or Tableau
+
+There's a high up-front cost and complexity to using a third-party tool. Additionally, there may be hidden constraints
+in such tools that we don't want to deal with while we're in an iterative stage of this effort.
+
+## Status
+
+Proposed
+
+## Consequences
+
+* We will have more nightly cron tasks to refresh the reporting views. We need to monitor the performance of these
```

Comment:
One more consequence to list here is that there will be a temptation to add business logic to the reporting views that often would be better located in the application itself.  For example, we have encoded the idea of `risk_level` into some mat views before as [hardcoded SQL](https://github.com/simpledotorg/simple-server/blob/13299d2bc34312f49eff15249864a8a2615c78c3/db/views/patient_summaries_v01.sql#L35-L48) that generates the views. You can imagine if this were something that stakeholders and product wanted to iterate on, it would be more difficult to change and validate as raw SQL as compared to ruby logic.

One other example that has come up is answering the question of "what protocol step should a patient be on?" ...I can imagine a 20 line SQL query that can answer that question...but I don't think I'd want to write it or maintain it, at least not to start. 😁  That is the sort of 'core business logic' that I think belongs squarely in the ruby domain, at least for awhile while we iron out the rough edges and make sure it matches the health care folks want.

There is a tradeoff between encoding things into SQL as compared to Ruby ...Ruby is easier to change and iterate on, whereas SQL can consolidated and reduce duplication, but is often harder to change and test. Something worth considering and pointing out here.